src/com/ms/silverking/numeric/Histogram.java

```diff
@@ -9,275 +9,273 @@
+
+  private static final double NOT_LOG_SCALE = 0.0;
+
+  public Histogram(double min, double max, int numBins) {
+    this(min, max, numBins, NOT_LOG_SCALE);
+  }
+
+  public Histogram(double min, double max, int numBins, double logBase) {
+    this.min = min;
+    this.max = max;
+    this.numBins = numBins;
+    bins = new Bin[numBins];
+    this.logBase = logBase;
+    if (logBase == NOT_LOG_SCALE) {
+      binSize = (max - min) / (double) numBins;
+      log = null;
+      initBins();
+    } else {
+      binSize = -1;
+      log = initLogBins();
     }
-    
-    public boolean logScale() {
-        return logBase != NOT_LOG_SCALE;
+  }
+
+  private void initBins() {
+    double binMin;
+
+    binMin = 0.0;
+    for (int i = 0; i < numBins; i++) {
+      double binMax;
+
+      if (i < numBins - 1) {
+        binMax = binMin + binSize;
+      } else {
+        binMax = max;
+      }
+      bins[i] = new Bin(binMin, binMax);
+      binMin = binMax;
     }
-    
-    public void add(double value, int count) {
-        bins[binIndex(value)].add(count);
+  }
+
+  private LogBaseN initLogBins() {
+    double b;
+
+    b = Math.exp(Math.log(max - min) / (double) numBins);
+    for (int i = 0; i < numBins; i++) {
+      double binMin;
+      double binMax;
+
+      binMin = Math.pow(b, i);
+      if (i < numBins - 1) {
+        binMax = Math.pow(b, i + 1);
+      } else {
+        binMax = max;
+      }
+      bins[i] = new Bin(binMin, binMax);
+      binMin = binMax;
     }
-    
-    public void increment(double value) {
-        bins[binIndex(value)].increment();
+    return new LogBaseN(b);
+  }
+
+  public boolean logScale() {
+    return logBase != NOT_LOG_SCALE;
+  }
+
+  public void add(double value, int count) {
+    bins[binIndex(value)].add(count);
+  }
+
+  public void increment(double value) {
+    bins[binIndex(value)].increment();
+  }
+
+  private int binIndex(double value) {
+    if (value < min || value > max) {
+      throw new RuntimeException("value out of range: " + value);
     }
-    
-    private int binIndex(double value) {
-        if (value < min || value > max) {
-            throw new RuntimeException("value out of range: "+ value);
+    if (value == max) {
+      return numBins - 1;
+    } else {
+      if (logBase == NOT_LOG_SCALE) {
+        return (int) ((value - min) / binSize);
+      } else {
+        if (value == min) {
+          return 0;
+        } else {
+          //System.out.println(log +" "+ value +" "+ min + log.log(value - min));
+          return (int) log.log(value - min);
         }
-        if (value == max) {
-            return numBins - 1;
+      }
+    }
+  }
+
+  public static Histogram parse(String fileName, int numBins) throws IOException {
+    return parse(new FileInputStream(fileName), numBins);
+  }
+
+  public static Histogram parse(String fileName, int numBins, double logBase) throws IOException {
+    return parse(new FileInputStream(fileName), numBins, logBase);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor call: *FileInputStream*. Currently, there are execution paths that do not contain closure statements, for example, when Histogram.parse throws an exception. Either a) close the object returned by *FileInputStream()* in a try-finally block or b) close the resource by declaring the object returned by *FileInputStream()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).